packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1468 @@
+     */
+    function previewKghUndelegate(
+        address validator,
+        uint256 tokenId
+    ) public view virtual returns (uint128) {
+        return
+            _convertToKghAssets(
+                validator,
+                _vault[validator].kghDelegators[msg.sender].kghShares[tokenId]
+            );
+    }
+
+    /**
+     * @notice Returns the total amount of KRO assets held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of KRO assets held by the vault.
+     */
+    function totalKroAssets(address validator) public view virtual returns (uint128) {
+        return _vault[validator].asset.totalKro;
+    }
+
+    /**
+     * @notice Returns the total number of KGHs held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total number of KGHs held by the vault.
+     */
+    function totalKghNum(address validator) public view virtual returns (uint128) {
+        return _vault[validator].asset.totalKgh;
+    }
+
+    /**
+     * @notice Returns the total amount of KGH assets held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of KGH assets held by the vault.
+     */
+    function _totalKghAssets(address validator) internal view virtual returns (uint128) {
+        Vault storage vault = _vault[validator];
+        return vault.asset.totalKgh * VKRO_PER_KGH + vault.reward.boostedReward;
+    }
+
+    /**
+     * @notice Returns the total amount of KRO shares held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of shares held by the validator vault.
+     */
+    function _totalKroShares(address validator) internal view virtual returns (uint128) {
+        return _vault[validator].asset.totalKroShares;
+    }
+
+    /**
+     * @notice Returns the total amount of KGH shares held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of shares held by the validator vault.
+     */
+    function _totalKghShares(address validator) internal view virtual returns (uint128) {
+        return _vault[validator].asset.totalKghShares;
+    }
+
+    /**
+     * @notice Delegate KRO to the validator and returns the amount of shares that the vault would
+     *         exchange.
+     *
+     * @param validator Address of the validator.
+     * @param assets    The amount of KRO to delegate.
+     *
+     * @return The amount of shares that the Vault would exchange for the amount of assets provided.
+     */
+    function delegate(
+        address validator,
+        uint128 assets
+    ) external checkIsActive(validator) returns (uint128) {
+        require(
+            assets <= ASSET_TOKEN.balanceOf(msg.sender),
+            "AssetManager: Exceeds max delegate amount"
+        );
```

Comment:
This would be checked in `ERC20.transferFrom` function, so I think it is unnecessary.

Reply:
I'll fix it, thanks :)